src/core/endpoint.ml

```diff
@@ -0,0 +1,39 @@
+type t = Domain of [ `host ] Domain_name.t | IP of Ipaddr.t
```

Comment:
How about expanding this to be `IPv4 of Ipaddr.V4.t | IPv6 of Ipaddr.V6.t` to save some indirection? There's unlikely to be any more IP variants in the next decade :-)

Reply:
what would the advantage be? I'm in favour of `IP of Ipaddr.t`.

Reply:
just a minor improvement in that it saves a boxed value. Probably not significant, except that a resolver can come back with multiple IP addresses for an endpoint.  It's a potential problem in the current `resolver` interface that only one `Endpoint.t` can come back. What should do the selection between an `A` and `AAAA` record for the same DNS entry?

Reply:
ic. I have not in-depth looked into the current resolver interface, but how to deal with multiple addresses and address families is nicely documented in "happy eyeballs v2" https://tools.ietf.org/html/rfc8305 -- I'd suggest to keep that out of this PR though (but eventually discuss how to deal with it before the 3.0 release?)

Reply:
I think the problem about many endpoints is a good question but, as @hannesm said, we should make a new issue/PR about that.

Reply:
that's fine by me - would you be able to create that followup issue @dinosaure? I've not read rfc8305 in detail yet but hope to do so early next week

Reply:
yes, I will start to clean issues today :+1: